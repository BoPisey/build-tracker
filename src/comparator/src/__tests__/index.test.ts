import BuildComparator from '..';
import Build from '@build-tracker/build';

const build1 = new Build({ revision: '1234567', timestamp: 1234567 }, [
  { name: 'churros', hash: 'abc', sizes: { stat: 456, gzip: 90 } },
  { name: 'tacos', hash: 'abc', sizes: { stat: 123, gzip: 45 } }
]);

const build2 = new Build({ revision: '8901234', timestamp: 8901234 }, [
  { name: 'burritos', hash: 'def', sizes: { stat: 469, gzip: 93 } },
  { name: 'tacos', hash: 'abc', sizes: { stat: 123, gzip: 43 } }
]);

const artifactFilters = [/burritos/, /churros/];

describe('BuildComparator', () => {
  describe('artifactFilters', () => {
    const comparator = new BuildComparator({ builds: [build1, build2], artifactFilters });

    test('filters artifact sizes from the totals', () => {
      expect(comparator.matrixTotal).toMatchSnapshot();
    });

    test('filters artifacts from artifactNames', () => {
      expect(comparator.artifactNames).toEqual(['tacos']);
    });
  });

  describe('artifactNames', () => {
    test('includes all artifact names', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.artifactNames).toEqual(expect.arrayContaining(['burritos', 'churros', 'tacos']));
    });
  });

  describe('buildDeltas', () => {
    test('are cached across gets', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.buildDeltas).toBe(comparator.buildDeltas);
    });
  });

  describe('matrix', () => {
    test('includes a header', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.matrix.header).toEqual(comparator.matrixHeader);
      expect(comparator.matrix.header).toMatchSnapshot();
    });

    test('includes a total', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.matrix.total).toEqual(comparator.matrixTotal);
      expect(comparator.matrix.total).toMatchSnapshot();
    });

    test('includes a body of all artifacts', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.matrix.body).toEqual(comparator.matrixBody);
      expect(comparator.matrix.body).toMatchSnapshot();
    });

    test('does not include filtered artifacts', () => {
      const comparator = new BuildComparator({ builds: [build1, build2], artifactFilters });
      expect(comparator.matrix.body).toHaveLength(1);
      // @ts-ignore
      expect(comparator.matrix.body[0][0].text).toEqual('tacos');
    });
  });

  describe('getSum', () => {
    test('gets a row of sums', () => {
      const comparator = new BuildComparator({ builds: [build1, build2], artifactFilters });
      const sum = comparator.getSum(['churros', 'tacos']);
      expect(sum).toMatchSnapshot();
    });

    test('filters on exact names', () => {
      const comparator = new BuildComparator({
        builds: [
          new Build({ revision: '1234567', timestamp: 1234567 }, [
            { name: 'i18n/en', hash: 'abc', sizes: { stat: 456, gzip: 90 } },
            { name: 'i18n/en-GB', hash: 'abc', sizes: { stat: 123, gzip: 45 } }
          ])
        ]
      });
      const sum = comparator.getSum(['i18n/en']);
      expect(sum).toMatchSnapshot();
    });
  });

  describe('getAscii', () => {
    test('gets an ASCII table', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.getAscii()).toMatchSnapshot();
    });

    test('can filter rows', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      const rowFilter = (row): boolean => {
        return row.some(cell => {
          if (cell.sizes && 'gzip' in cell.sizes) {
            return cell.sizes.gzip > 50;
          }
          return false;
        });
      };
      expect(comparator.getAscii({ rowFilter })).toMatchSnapshot();
    });

    test('does not include filtered artifacts', () => {
      const comparator = new BuildComparator({ builds: [build1, build2], artifactFilters });
      expect(comparator.getAscii()).toMatchSnapshot();
    });
  });

  describe('getCsv', () => {
    test('gets a CSV formatted table', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      expect(comparator.getCsv()).toMatchSnapshot();
    });

    test('can filter rows', () => {
      const comparator = new BuildComparator({ builds: [build1, build2] });
      const rowFilter = (row): boolean => {
        return row.some(cell => {
          if (cell.sizes && 'gzip' in cell.sizes) {
            return cell.sizes.gzip > 50;
          }
          return false;
        });
      };
      expect(comparator.getCsv({ rowFilter })).toMatchSnapshot();
    });

    test('does not include filtered artifacts', () => {
      const comparator = new BuildComparator({ builds: [build1, build2], artifactFilters });
      expect(comparator.getCsv()).toMatchSnapshot();
    });
  });
});
