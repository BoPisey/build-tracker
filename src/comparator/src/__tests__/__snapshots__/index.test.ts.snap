// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildComparator artifactFilters filters artifact sizes from the totals 1`] = `
Array [
  Object {
    "text": "All",
    "type": "artifact",
  },
  Object {
    "gzip": 45,
    "stat": 123,
    "type": "total",
  },
  Object {
    "gzip": 43,
    "stat": 123,
    "type": "total",
  },
  Object {
    "againstRevision": "1234567",
    "sizes": Object {
      "gzip": -2,
      "gzipPercent": -0.044444444444444446,
      "stat": 0,
      "statPercent": 0,
    },
    "type": "totalDelta",
  },
]
`;

exports[`BuildComparator getAscii can filter rows 1`] = `
".----------------------------------------------.
|          | 1234567 | 8901234 |      Δ1       |
|----------|---------|---------|---------------|
| burritos | 0       | 93      | 93 (100.0%)   |
| churros  | 90      | 0       | -90 (-100.0%) |
'----------------------------------------------'"
`;

exports[`BuildComparator getAscii does not include filtered artifacts 1`] = `
".----------------------------------------.
|       | 1234567 | 8901234 |     Δ1     |
|-------|---------|---------|------------|
| tacos | 45      | 43      | -2 (-4.4%) |
'----------------------------------------'"
`;

exports[`BuildComparator getAscii gets an ASCII table 1`] = `
".----------------------------------------------.
|          | 1234567 | 8901234 |      Δ1       |
|----------|---------|---------|---------------|
| burritos | 0       | 93      | 93 (100.0%)   |
| churros  | 90      | 0       | -90 (-100.0%) |
| tacos    | 45      | 43      | -2 (-4.4%)    |
'----------------------------------------------'"
`;

exports[`BuildComparator getCsv can filter rows 1`] = `
",1234567,8901234,Δ1
burritos,0,93,93 (100.0%)
churros,90,0,-90 (-100.0%)"
`;

exports[`BuildComparator getCsv does not include filtered artifacts 1`] = `
",1234567,8901234,Δ1
tacos,45,43,-2 (-4.4%)"
`;

exports[`BuildComparator getCsv gets a CSV formatted table 1`] = `
",1234567,8901234,Δ1
burritos,0,93,93 (100.0%)
churros,90,0,-90 (-100.0%)
tacos,45,43,-2 (-4.4%)"
`;

exports[`BuildComparator getSum filters on exact names 1`] = `
Array [
  Object {
    "text": "Sum",
    "type": "text",
  },
  Object {
    "sizes": Object {
      "gzip": 90,
      "stat": 456,
    },
    "type": "total",
  },
]
`;

exports[`BuildComparator getSum gets a row of sums 1`] = `
Array [
  Object {
    "text": "Sum",
    "type": "text",
  },
  Object {
    "sizes": Object {
      "gzip": 135,
      "stat": 579,
    },
    "type": "total",
  },
  Object {
    "againstRevision": "1234567",
    "sizes": Object {
      "gzip": -2,
      "gzipPercent": -0.044444444444444446,
      "stat": 0,
      "statPercent": 0,
    },
  },
  Object {
    "sizes": Object {
      "gzip": 43,
      "stat": 123,
    },
    "type": "total",
  },
  Object {
    "againstRevision": "1234567",
    "sizes": Object {
      "gzip": -2,
      "gzipPercent": -0.044444444444444446,
      "stat": 0,
      "statPercent": 0,
    },
  },
]
`;

exports[`BuildComparator matrix includes a body of all artifacts 1`] = `
Array [
  Array [
    Object {
      "text": "burritos",
      "type": "artifact",
    },
    Object {
      "hash": "",
      "name": "",
      "sizes": Object {},
      "type": "total",
    },
    Object {
      "hash": "def",
      "name": "burritos",
      "sizes": Object {
        "gzip": 93,
        "stat": 469,
      },
      "type": "total",
    },
    Object {
      "hashChanged": true,
      "name": "burritos",
      "sizes": Object {
        "gzip": 93,
        "stat": 469,
      },
      "type": "delta",
    },
  ],
  Array [
    Object {
      "text": "churros",
      "type": "artifact",
    },
    Object {
      "hash": "abc",
      "name": "churros",
      "sizes": Object {
        "gzip": 90,
        "stat": 456,
      },
      "type": "total",
    },
    Object {
      "hash": "",
      "name": "",
      "sizes": Object {},
      "type": "total",
    },
    Object {
      "hashChanged": true,
      "name": "churros",
      "sizes": Object {
        "gzip": -90,
        "gzipPercent": -1,
        "stat": -456,
        "statPercent": -1,
      },
      "type": "delta",
    },
  ],
  Array [
    Object {
      "text": "tacos",
      "type": "artifact",
    },
    Object {
      "hash": "abc",
      "name": "tacos",
      "sizes": Object {
        "gzip": 45,
        "stat": 123,
      },
      "type": "total",
    },
    Object {
      "hash": "abc",
      "name": "tacos",
      "sizes": Object {
        "gzip": 43,
        "stat": 123,
      },
      "type": "total",
    },
    Object {
      "hashChanged": false,
      "name": "tacos",
      "sizes": Object {
        "gzip": -2,
        "gzipPercent": -0.044444444444444446,
        "stat": 0,
        "statPercent": 0,
      },
      "type": "delta",
    },
  ],
]
`;

exports[`BuildComparator matrix includes a header 1`] = `
Array [
  Object {
    "text": "",
    "type": "text",
  },
  Object {
    "revision": "1234567",
    "type": "revision",
  },
  Object {
    "revision": "8901234",
    "type": "revision",
  },
  Object {
    "againstRevision": "8901234",
    "deltaIndex": 1,
    "revision": "8901234",
    "type": "revisionDelta",
  },
]
`;

exports[`BuildComparator matrix includes a total 1`] = `
Array [
  Object {
    "text": "All",
    "type": "artifact",
  },
  Object {
    "gzip": 135,
    "stat": 579,
    "type": "total",
  },
  Object {
    "gzip": 136,
    "stat": 592,
    "type": "total",
  },
  Object {
    "againstRevision": "1234567",
    "sizes": Object {
      "gzip": 1,
      "gzipPercent": 0.007407407407407408,
      "stat": 13,
      "statPercent": 0.022452504317789293,
    },
    "type": "totalDelta",
  },
]
`;
